File: Graphics/Engine/Components/ObjLoader.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Numerics;

namespace GraphicsLibrary
{
    public static class ObjLoader
    {
        public static Mesh Load(string filePath)
        {
            if (!File.Exists(filePath))
                throw new FileNotFoundException("OBJ file not found", filePath);

            var positions = new List<Vector3>();
            var normals = new List<Vector3>();
            var uvs = new List<Vector2>();
            var mesh = new Mesh();

            foreach (var line in File.ReadLines(filePath))
            {
                string l = line.Trim();
                if (string.IsNullOrWhiteSpace(l) || l.StartsWith("#"))
                    continue;

                var parts = l.Split(' ', StringSplitOptions.RemoveEmptyEntries);
                if (parts.Length == 0) continue;

                switch (parts[0])
                {
                    case "v":
                        positions.Add(new Vector3(
                            float.Parse(parts[1]),
                            float.Parse(parts[2]),
                            float.Parse(parts[3])
                        ));
                        break;

                    case "vn":
                        normals.Add(new Vector3(
                            float.Parse(parts[1]),
                            float.Parse(parts[2]),
                            float.Parse(parts[3])
                        ).Normalized());
                        break;

                    case "vt":
                        uvs.Add(new Vector2(
                            float.Parse(parts[1]),
                            float.Parse(parts[2])
                        ));
                        break;

                    case "f":
                        if (parts.Length < 4)
                            continue;

                        var faceIndices = new List<(int pos, int uv, int norm)>();
                        for (int i = 1; i < parts.Length; i++)
                        {
                            var idx = parts[i].Split('/');
                            int p = int.Parse(idx[0]) - 1;
                            int t = idx.Length > 1 && !string.IsNullOrEmpty(idx[1]) ? int.Parse(idx[1]) - 1 : -1;
                            int n = idx.Length > 2 && !string.IsNullOrEmpty(idx[2]) ? int.Parse(idx[2]) - 1 : -1;

                            faceIndices.Add((p, t, n));
                        }

                        for (int i = 1; i < faceIndices.Count - 1; i++)
                        {
                            mesh.AddTriangle(new Triangle(
                                CreateVertex(faceIndices[0], positions, uvs, normals),
                                CreateVertex(faceIndices[i], positions, uvs, normals),
                                CreateVertex(faceIndices[i + 1], positions, uvs, normals)
                            ));
                        }
                        break;
                }
            }

            return mesh;
        }

        private static Vertex CreateVertex((int pos, int uv, int norm) idx, List<Vector3> positions, List<Vector2> uvs, List<Vector3> normals)
        {
            Vector3 position = positions[idx.pos];
            Vector3 normal = idx.norm >= 0 && idx.norm < normals.Count ? normals[idx.norm] : Vector3.UnitY;
            Vector2 uv = idx.uv >= 0 && idx.uv < uvs.Count ? uvs[idx.uv] : Vector2.Zero;
            return new Vertex(position, normal, Color.White, uv);
        }
    }
}

File: Graphics_Native/Engine/Components/ObjLoader.cpp
#include "ObjLoader.h"
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <stdexcept>
#include <cstdlib>

static int ParseInt(const char* s) {
    // Use strtol instead of stoi — stops at non-digit chars (\r, etc.) without throwing
    char* end;
    long v = std::strtol(s, &end, 10);
    return (end == s) ? 0 : (int)v;
}

static void ParseIndex(const std::string& token,
                        int& posIdx, int& uvIdx, int& normIdx) {
    posIdx = uvIdx = normIdx = -1;
    const char* p = token.c_str();
    // Component 0: position
    char* end;
    long v0 = std::strtol(p, &end, 10);
    if (end == p) return;
    posIdx = (int)v0 - 1;
    if (*end != '/') return;
    p = end + 1;
    // Component 1: uv (may be empty, e.g. "1//2")
    if (*p != '/') {
        long v1 = std::strtol(p, &end, 10);
        if (end != p) uvIdx = (int)v1 - 1;
        p = end;
    }
    if (*p != '/') return;
    p++;
    // Component 2: normal
    long v2 = std::strtol(p, &end, 10);
    if (end != p) normIdx = (int)v2 - 1;
}

static Vertex MakeVertex(int posIdx, int uvIdx, int normIdx,
                          const std::vector<Vector3>& positions,
                          const std::vector<Vector2>& uvs,
                          const std::vector<Vector3>& normals) {
    Vertex v;
    // Bounds-check every index — bad data should never crash
    v.Position = (posIdx  >= 0 && posIdx  < (int)positions.size()) ? positions[posIdx]  : Vector3{};
    v.Normal   = (normIdx >= 0 && normIdx < (int)normals.size())   ? normals[normIdx]   : Vector3::UnitY;
    v.UV       = (uvIdx   >= 0 && uvIdx   < (int)uvs.size())       ? uvs[uvIdx]         : Vector2::Zero;
    v.Color    = Color::White;
    return v;
}

Mesh ObjLoader::Load(const std::string& filePath) {
    std::ifstream file(filePath);
    if (!file.is_open())
        throw std::runtime_error("OBJ file not found: " + filePath);

    std::vector<Vector3> positions;
    std::vector<Vector3> normals;
    std::vector<Vector2> uvs;
    Mesh mesh;

    // Reserve generous estimates to avoid repeated reallocs
    positions.reserve(8192);
    normals.reserve(8192);
    uvs.reserve(8192);
    mesh.Triangles.reserve(16384);

    // Fixed-size face buffer — handles up to 32-gon polygons without heap allocation
    Vertex faceVerts[32];

    std::string line;
    while (std::getline(file, line)) {
        if (line.empty()) continue;
        // Strip trailing \r for CRLF files
        if (line.back() == '\r') line.pop_back();
        if (line.empty() || line[0] == '#') continue;

        const char* lp = line.c_str();
        // Skip leading whitespace
        while (*lp == ' ' || *lp == '\t') ++lp;

        if (lp[0] == 'v' && lp[1] == ' ') {
            float x = 0, y = 0, z = 0;
            sscanf(lp + 2, "%f %f %f", &x, &y, &z);
            positions.push_back({ x, y, z });

        } else if (lp[0] == 'v' && lp[1] == 'n') {
            float x = 0, y = 0, z = 0;
            sscanf(lp + 3, "%f %f %f", &x, &y, &z);
            normals.push_back(Vector3{ x, y, z }.Normalized());

        } else if (lp[0] == 'v' && lp[1] == 't') {
            float u = 0, v = 0;
            sscanf(lp + 3, "%f %f", &u, &v);
            uvs.push_back({ u, v });

        } else if (lp[0] == 'f' && lp[1] == ' ') {
            std::istringstream ss(lp + 2);
            std::string token;
            int faceCount = 0;
            while (ss >> token && faceCount < 32) {
                int p, t, n;
                ParseIndex(token, p, t, n);
                if (p < 0 || p >= (int)positions.size()) continue;
                faceVerts[faceCount++] = MakeVertex(p, t, n, positions, uvs, normals);
            }
            for (int i = 1; i < faceCount - 1; i++)
                mesh.Triangles.push_back({ faceVerts[0], faceVerts[i], faceVerts[i+1] });
        }
    }

    return mesh;
}
